### PR TITLE
feat: Implement a k-dimensional range class

### PR DESCRIPTION
**Pull request by stephenswat**
This pull request is number two in a series of four in my attempt to implement an orthogonal range search seed finder. Part one is #985.

The new class represents an axis-aligned (thereby orthogonal) hyperrectangle in a space of arbitrary dimension. This can be achieved quite elegantly by simply taking the product of a number of one-dimensional intervals.

I've also included some helper methods which will be useful in the implementation of the k-d tree.

As with the other pull requests in this series, this also adds tests for this new class.

This cannot be merged until #985 is merged.

### CODE REVIEW

File: Core/include/Acts/Utilities/RangeXD.hpp [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+
+    return true;
+  }
+
+  /// @brief Access one of the dimensional ranges of the volume
+  ///
+  /// @param i The index of the dimension to access
+  /// @return A reference to the dimension contained in this range
+  Range1D<Type>& operator[](const std::size_t& i) {
+    return m_dims[static_cast<std::size_t>(i)];
```
Comment (asalzburger) [inline Core/include/Acts/Utilities/RangeXD.hpp:73]: Is that static_cast here necessary?
Looks like you have enforced the typ already.

File: Core/include/Acts/Utilities/RangeXD.hpp [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+  }
+
+  /// @brief Access one of the dimensional ranges of the volume
+  ///
+  /// @param i The index of the dimension to access
+  /// @return A reference to the dimension contained in this range
+  const Range1D<Type>& operator[](const std::size_t& i) const {
+    return m_dims[static_cast<std::size_t>(i)];
```
Comment (asalzburger) [inline Core/include/Acts/Utilities/RangeXD.hpp:81]: Same as above.

File: Core/include/Acts/Utilities/RangeXD.hpp [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+  /// of their n dimensions.
+  ///
+  /// @param o The other range to check for equality
+  ///
+  /// @return true The ranges are equal
+  /// @return false The ranges are not equal
+  bool operator==(const RangeXD<Dims, Type, Vector>& o) const {
+    for (std::size_t i = 0; i < Dims; ++i) {
+      if (!(this->m_dims[i] == o[i])) {
+        return false;
```
Comment (asalzburger) [inline Core/include/Acts/Utilities/RangeXD.hpp:96]: I suppose you could rely on the fact that the `Vector` typename implements an `==` operator, but also ok this way.

Comment (asalzburger) [reply Core/include/Acts/Utilities/RangeXD.hpp:96]: A sorry, my mistake - you rely on this already, this concerns the dimensions.

File: Core/include/Acts/Utilities/RangeXD.hpp [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+    return true;
+  }
+
+  /// @brief Represent the range as a string
+  ///
+  /// This method produces a helpful string that can be used to debug the range
+  /// if needed. Not really designed to be used in production code.
+  ///
+  /// @return A string representing the range
+  std::string str(void) const {
```
Comment (asalzburger) [inline Core/include/Acts/Utilities/RangeXD.hpp:212]: We have this at other places, can you please check if this is called `str(void)` or (what I have in mind) `toString(void)`?

I just want to make sure we keep similar naming conventions.

**Review by asalzburger — CHANGES_REQUESTED** [review]
Straight-forwarded multidimensional extension of the `Range1D` implementation - I would just ask to cross-check if the `str()` to string method is coherently named as similar ones in the code base.

File: Core/include/Acts/Utilities/RangeXD.hpp (hunk 4 above)
Comment (stephenswat) [reply Core/include/Acts/Utilities/RangeXD.hpp:212]: Indeed it seems it's `toString(void)` in most places in the Acts code. I'll change it!

File: Core/include/Acts/Utilities/RangeXD.hpp (hunk 1 above)
Comment (stephenswat) [reply Core/include/Acts/Utilities/RangeXD.hpp:73]: Well spotted, this is a vestige of some older design that can be removed.

**Comment by stephenswat** [comment]
Changes incorporated, re-running CI.

**Review by asalzburger — APPROVED** [review]

**Comment by paulgessinger** [comment]
I know it's maybe too late now since this is already merged, but I'm not a huge fan of these operator overloads. 
Could you consider using named functions to implement this functionality?

**Comment by stephenswat** [comment]
Sure thing! I think this is a question of what you're used to - I borrowed this notation from Python sets which use the ampersand for intersections. I am more than happy to replace this with a named method, though.

**Pull request merged**